docs/adrs/0003-profile-layering.md

```diff
@@ -0,0 +1,119 @@
+# 3. Profile Layering
+
+Date: 2021-11-16
+
+## Status
+
+Accepted
+
+## Context
+
+There are cases where we require to install profiles in a specific order,
+perhaps the components being installed are required to be functional before
+further workloads are installed into the cluster.
+
+Flux HelmReleases support this through the [`dependsOn` mechanism](https://fluxcd.io/docs/components/helm/helmreleases/#helmrelease-dependencies).
+
+Because the `HelmReleases` are generated inside the CAPI service, we want to
+provide a way to configure this dependency without having to manually edit the
+files.
+
+We may not always know the specific dependency to be able to set an explicit
+dependency, for example, "certificate management" should be in place before
+"service mesh".
+
+When installing a set of profiles, it should be possible to declaratively say
+that some profiles should be installed before others if this is required.
+
+## Decision
+
+Support Helm Chart [annotation](https://helm.sh/docs/topics/charts/#the-chartyaml-file) that indicates a layer-like ordering for unrelated profiles.
+
+The `weave.works/layer` indicates that a Profile chart should be applied in a
+specific layer.
+
+For example, with the following two Profile charts.
+
+
+```yaml
+apiVersion: v2
+name: base-profile
+version: 1.0.3
+annotations:
+  weave.works/profile: base
+  weave.works/layer: layer-0
+```
+
+```yaml
+apiVersion: v2
+name: demo-profile
+version: 0.1.1
+annotations:
+  weave.works/profile: demo
+  weave.works/layer: layer-1
+```
+
+This would result in HelmReleases that looked like this:
+
+
+```yaml
+apiVersion: helm.toolkit.fluxcd.io/v2beta1
+kind: HelmRelease
+metadata:
+  name: base-profile-default
+spec:
+  chart:
+    spec:
+      chart: base-profile
+      version: "1.0.3"
+      sourceRef:
+        kind: HelmRepository
+        name: demo
+```
+
+With the dependent HelmRelease:
+
+```yaml
+apiVersion: helm.toolkit.fluxcd.io/v2beta1
+kind: HelmRelease
+metadata:
+  name: demo-profile-default
+  namespace: default
+spec:
+  chart:
+    spec:
+      chart: demo-profile
+      version: "0.1.1"
+      sourceRef:
+        kind: HelmRepository
+        name: demo
+  dependsOn:
+    - name: base-profile-default
+```
+
+The layers are sorted lexicographically, which provides flexibility in naming,
```

Comment:
I'm not following this one. Why should / would we sort layers?

Reply:
Because we might want to have layers like `infrastructure` and `web` or something (there's no standard _yet_)

Reply:
But what would sorting do in that regard? Sorry, I'm just not seeing where or what you mean by sorting here. :) Sort what, how and where? And where would the sorted result be at?